File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
   let buyPriceIndex = 0;
   let sellPriceIndex = buyPriceIndex + 1;
   let profit = 0;
   let n = prices.length;
   while(sellPriceIndex < n) {
    if(prices[buyPriceIndex] <= prices[sellPriceIndex]) {
        let diff = prices[sellPriceIndex] - prices[buyPriceIndex];
        if(diff > profit) profit = diff; // maximizing profit
        // buyPriceIndex++;
        sellPriceIndex++;
    } else if(prices[buyPriceIndex] > prices[sellPriceIndex]) {
        buyPriceIndex = sellPriceIndex;
        sellPriceIndex++;
    } // using if , elseif as buy and sell are happening only when buy happens not before that.
   }

   return profit;
};

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices.length <= 1) return 0;
        int min = 10000;
        int minIndex = 0;
        int max = 0;
        int maxIndex = 0;
        int maxProfit = 0;
        for(int i=0; i<prices.length; i++) {
            if(prices[i] < min) {
                min = prices[i];
                minIndex = i;
                max = min;
                maxIndex = i;
            } else if(prices[i] > max) {
                max = prices[i];
                maxIndex = i;
            }

            if(maxIndex > minIndex) {
                int diff = max - min;
                if(diff > maxProfit) {
                    maxProfit = max - min;
                }
            }
        }

        return maxProfit;
    }
}